lib/batch_verification/src/server.rs

```diff
@@ -0,0 +1,168 @@
+use crate::{
+    BATCH_VERIFICATION_WIRE_FORMAT_VERSION, BatchVerificationRequest,
+    BatchVerificationRequestCodec, BatchVerificationResponse, BatchVerificationResponseDecoder,
+};
+use futures::{SinkExt, StreamExt};
+use tokio::io::BufReader;
+use tokio::net::ToSocketAddrs;
+use tokio::sync::broadcast;
+use tokio::sync::mpsc;
+use tokio::{
+    io::AsyncWriteExt,
+    net::{TcpListener, TcpStream},
+};
+use tokio_util::codec::{FramedRead, FramedWrite};
+use zksync_os_l1_sender::batcher_model::BatchForSigning;
+use zksync_os_socket::skip_http_headers;
+
+/// Accepts connections from batch verification clients. Crafts and sends
+/// BatchVerificationRequests to all clients. Receives responses and forwards
+/// them through the channel to batch_response_processor
+pub struct BatchVerificationServer {
+    verification_request_broadcast: broadcast::Sender<BatchVerificationRequest>,
+    response_sender: mpsc::Sender<BatchVerificationResponse>,
+}
+
+#[derive(Debug, thiserror::Error)]
+#[allow(clippy::large_enum_variant)]
+pub enum BatchVerificationRequestError {
+    #[error("Not enough clients connected")]
+    NotEnoughClients,
+    #[error("Failed to send batch verification request: {0}")]
+    SendError(#[from] broadcast::error::SendError<BatchVerificationRequest>),
+}
+
+impl BatchVerificationServer {
+    pub fn new() -> (Self, mpsc::Receiver<BatchVerificationResponse>) {
+        let (response_sender, response_receiver) = mpsc::channel(100);
+        let (verification_request_broadcast, _rx_unused) = broadcast::channel(16);
+
+        let server = Self {
+            verification_request_broadcast,
+            response_sender,
+        };
+
+        (server, response_receiver)
+    }
+
+    /// Start the TCP server that accepts connections from external nodes
+    pub async fn run_server(&self, address: impl ToSocketAddrs) -> anyhow::Result<()> {
+        let listener = TcpListener::bind(address).await?;
+        let response_sender = self.response_sender.clone();
+
+        loop {
+            let (socket, addr) = listener.accept().await?;
+            let verification_request_rx = self.verification_request_broadcast.subscribe();
+            let response_sender = response_sender.clone();
+            let client_addr = addr.to_string();
+
+            tokio::spawn(async move {
+                if let Err(e) = Self::handle_client(
+                    socket,
+                    client_addr,
+                    verification_request_rx,
+                    response_sender,
+                )
+                .await
+                {
+                    tracing::info!("Error handling client {}: {}", addr, e);
+                }
+            });
+        }
+    }
+
+    async fn handle_client(
```

Comment:
1. Shouldn't we have some kind of handshake to ensure that both parties use the same version of the protocol? Right now, it looks like if client doesn't support required version of the protocol, the only option it has is to close the connection (which is not very helpful).
2. Don't we want the communication to be encrypted? E.g. during the handshake we verify that the person who connects is among `accepted_signers`, and then derive a shared ephemeral secret to encode the communication (similarly to what rlpx or noise protocol does). Otherwise, it looks like we're very prone to any kind of MITM attack where the messages in either direction can be altered (causing valid blocks to be not signed or returning malformed signatures to the server).